File: javascript/src/aoc_2022/day_04.mjs
export const dayNumber = 4;

function elfRange(elf) {
  const [s, e] = elf.split("-");
  return [parseInt(s, 10), parseInt(e, 10)];
}

function contains([as, ae], [bs, be]) {
  return as <= bs && ae >= be;
}

function inside(as, [bs, be]) {
  return bs <= as && be >= as;
}

export function part1(input) {
  let count = 0;
  for (const line of input.trim().split("\n")) {
    const [l, r] = line.split(",");
    const lr = elfRange(l);
    const rr = elfRange(r);
    if (contains(lr, rr) || contains(rr, lr)) {
      count++;
    }
  }
  return count;
}

export function part2(input) {
  let count = 0;
  for (const line of input.trim().split("\n")) {
    const [l, r] = line.split(",");
    const lr = elfRange(l);
    const rr = elfRange(r);
    if (inside(lr[0], rr) || inside(rr[0], lr)) {
      count++;
    }
  }
  return count;
}

File: go/aoc_2023/internal/day04/day04.go
package day04

import (
	"bufio"
	"fmt"
	"gitlab.com/danscott/adventofcode/go/common/runner"
	"strconv"
)

type day04 struct{}

func (d *day04) Year() uint {
	return 2023
}

func (d *day04) Day() uint {
	return 4
}

func (d *day04) Part1(input *bufio.Scanner) string {
	total := 0
	input.Split(bufio.ScanWords)
	var wins map[int]interface{}
	score := 0
	inWins := true
	for input.Scan() {
		word := input.Text()
		switch word {
		case "Card":
			total += score
			wins = make(map[int]interface{})
			input.Scan()
			inWins = true
		case "|":
			inWins = false
			score = 0
		default:
			val, _ := strconv.ParseInt(word, 10, 32)
			if !inWins {
				if _, ok := wins[int(val)]; ok {
					if score == 0 {
						score = 1
					} else {
						score *= 2
					}
				}
			} else {
				wins[int(val)] = nil
			}
		}
	}
	return fmt.Sprint(total)
}

func (d *day04) Part2(input *bufio.Scanner) string {
	input.Split(bufio.ScanWords)
	cardCounts := make(map[int]int)
	current := 0
	var wins map[int]interface{}
	inWins := true
	winCt := 0
	for input.Scan() {
		word := input.Text()
		switch word {
		case "Card":
			if winCt > 0 {
				for i := current + 1; i <= current+winCt; i++ {
					cardCounts[i] += cardCounts[current]
				}
			}
			input.Scan()
			cardNumStr := input.Text()
			cardNumStr = cardNumStr[:len(cardNumStr)-1]
			cn, _ := strconv.ParseInt(cardNumStr, 10, 32)
			current = int(cn)
			cardCounts[current] += 1
			wins = make(map[int]interface{})
			inWins = true
		case "|":
			inWins = false
			winCt = 0

		default:
			val, _ := strconv.ParseInt(word, 10, 32)
			if !inWins {
				if _, ok := wins[int(val)]; ok {
					winCt += 1
				}
			} else {
				wins[int(val)] = nil
			}
		}
	}
	count := 0
	for i, c := range cardCounts {
		if i > current {
			break
		}
		count += c
	}
	return fmt.Sprint(count)
}

func New() runner.Day {
	return &day04{}
}
